### PR TITLE
Remove inter-package convenience imports

### PR DESCRIPTION
**Pull request by jonashen**
Fixes: #71.

### CODE REVIEW

File: garage/replay_buffer/replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+"""
```
Comment (gonzaiva) [inline garage/replay_buffer/replay_buffer.py]: Wasn't the replay buffer already added?

Comment (jonashen) [reply garage/replay_buffer/replay_buffer.py]: yes, must've been added in when I rebased?

File: garage/algos/__init__.py [hunk 2]
```diff
@@ -1,12 +1,8 @@
 from garage.algos.base import Algorithm
 from garage.algos.base import RLAlgorithm
 from garage.algos.batch_polopt import BatchPolopt
-from garage.algos.batch_polopt import BatchSampler
```
Comment (ryanjulian) [inline garage/algos/__init__.py]: Can you move BatchSampler into garage.sampler?

File: garage/envs/__init__.py [hunk 3]
```diff
@@ -3,10 +3,7 @@
 from garage.envs.env_spec import EnvSpec
 from garage.envs.grid_world_env import GridWorldEnv
 from garage.envs.identification_env import IdentificationEnv  # noqa: I100
-from garage.envs.noisy_env import DelayedActionEnv
-from garage.envs.noisy_env import NoisyObservationEnv
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply garage/envs/__init__.py]: you can alternatively move them into different files, but in Python it is quite common to have more than one class per file (module).

File: garage/envs/mujoco/randomization/__init__.py [hunk 4]
```diff
@@ -1,6 +1,4 @@
 from garage.envs.mujoco.randomization.randomized_env import randomize
-from garage.envs.mujoco.randomization.variation import Distribution
-from garage.envs.mujoco.randomization.variation import Method
```
Comment (ryanjulian) [inline garage/envs/mujoco/randomization/__init__.py]: I think these are okay. It means that Distribution and Method are both part of the public interface to `garage.envs.mujoco.randomization`.

File: garage/sampler/__init__.py [hunk 5]
```diff
@@ -1,8 +1,3 @@
-from garage.sampler.stateful_pool import SharedGlobal
 from garage.sampler.stateful_pool import singleton_pool
-from garage.sampler.parallel_sampler import truncate_paths  # noqa: I100
-from garage.sampler.stateful_pool import ProgBarCounter
 
-__all__ = [
-    "SharedGlobal", "singleton_pool", "truncate_paths", "ProgBarCounter"
```
Comment (ryanjulian) [inline garage/sampler/__init__.py]: This module is in general a mess, so there are no great decisions here. This looks okay.

Why are the base classes not imported?

File: garage/tf/core/__init__.py [hunk 6]
```diff
@@ -1,19 +1,11 @@
...
-from garage.tf.core.parameterized import JointParameterized
 
 __all__ = [
     "layers",
     "Parameterized",
     "LayersPowered",
-    "ConvNetwork",
-    "GRUNetwork",
-    "LSTMNetwork",
     "MLP",
```
Comment (ryanjulian) [inline garage/tf/core/__init__.py]: Why does MLP get to stay?

File: garage/tf/core/__init__.py [hunk 7]
```diff
@@ -1,19 +1,11 @@
...
 from garage.tf.core.parameterized import Parameterized
 from garage.tf.core.layers_powered import LayersPowered  # noqa: I100
-from garage.tf.core.network import ConvNetwork
-from garage.tf.core.network import GRUNetwork
-from garage.tf.core.network import LSTMNetwork
 from garage.tf.core.network import MLP
-from garage.tf.core.parameterized import JointParameterized
 
 __all__ = [
     "layers",
```
Comment (ryanjulian) [inline garage/tf/core/__init__.py]: Why does layers get to stay? It's a module, not a class.

File: garage/tf/core/__init__.py [hunk 8]
```diff
@@ -1,19 +1,11 @@
 from garage.tf.core import layers
 from garage.tf.core.parameterized import Parameterized
 from garage.tf.core.layers_powered import LayersPowered  # noqa: I100
-from garage.tf.core.network import ConvNetwork
```
Comment (ryanjulian) [inline garage/tf/core/__init__.py]: Another package which is a complete mess.

File: garage/tf/samplers/on_policy_vectorized_sampler.py [hunk 9]
```diff
@@ -6,7 +6,7 @@
 from garage.misc import tensor_utils
 import garage.misc.logger as logger
 from garage.misc.overrides import overrides
-from garage.sampler import ProgBarCounter
```
Comment (ryanjulian) [inline garage/tf/samplers/on_policy_vectorized_sampler.py]: Can you put `ProgBarCounter` into a util somewhere in misc? It doesn't even belong in sampler.

File: garage/theano/core/__init__.py [hunk 10]
```diff
@@ -1,12 +1,4 @@
...
 from garage.theano.core.lasagne_powered import LasagnePowered
-from garage.theano.core.network import ConvNetwork
-from garage.theano.core.network import GRUNetwork
 from garage.theano.core.network import MLP
 
-__all__ = [
-    "batch_norm", "OpLayer", "ParamLayer", "LasagnePowered", "ConvNetwork",
-    "GRUNetwork", "MLP"
-]
+__all__ = ["LasagnePowered", "MLP"]
```
Comment (ryanjulian) [inline garage/theano/core/__init__.py]: What makes MLP special? It's just another network.

File: garage/sampler/__init__.py [hunk 11]
```diff
@@ -1,8 +1,3 @@
-from garage.sampler.stateful_pool import SharedGlobal
 from garage.sampler.stateful_pool import singleton_pool
-from garage.sampler.parallel_sampler import truncate_paths  # noqa: I100
```
Comment (ryanjulian) [inline garage/sampler/__init__.py]: Can you move truncate paths into a util in sampler? It hardly belongs inside parallel_sampler.

Comment (jonashen) [reply garage/sampler/__init__.py]: done

File: garage/tf/samplers/on_policy_vectorized_sampler.py (hunk 9 above)
Comment (jonashen) [reply garage/tf/samplers/on_policy_vectorized_sampler.py]: done

File: garage/sampler/__init__.py (hunk 5 above)
Comment (jonashen) [reply garage/sampler/__init__.py]: done

File: garage/algos/__init__.py (hunk 2 above)
Comment (jonashen) [reply garage/algos/__init__.py]: done

File: garage/misc/counter.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
+"""A progress bar counter class."""
```
Comment (ryanjulian) [inline garage/misc/counter.py]: Call this file `prog_bar_counter.py` to keep with convention.

File: garage/sampler/base.py [hunk 13]
```diff
@@ -195,3 +196,33 @@ def process_samples(self, itr, paths):
         logger.record_tabular('MinReturn', np.min(undiscounted_returns))
 
         return samples_data
+
+
+class BatchSampler(BaseSampler):
```
Comment (ryanjulian) [inline garage/sampler/base.py]: Please put this in `batch_sampler.py` to keep with convention.

**Pull request merged**